### PR TITLE
NLog: Improved default JsonSerializerSettings, and reduced allocations

### PR DESCRIPTION
**Pull request by snakefoot**
- Convert enums to their string-value (instead of integer value)
- Do not throw exceptions on cyclic reference loops
- Skip JToken code-path for simple types that converts nicely to Protobuf
- Reuse the JsonSerializer instead of allocating new for every property

Still missing the ability to not serialize the whole world to an infinite depth.

### CODE REVIEW

**Comment by chrisdunelm** [comment]
Thanks for the PR. I'm currently on vacation, so will look at this in detail then I return in a week or so.

File: apis/Google.Cloud.Logging.NLog/Google.Cloud.Logging.NLog.Tests/GoogleStackdriverTargetTest.cs [hunk 1]
```diff
@@ -369,6 +378,30 @@ public async Task SingleLogEntryWithJsonCollectionProperties()
...
+                {
+                    googleTarget.SendJsonPayload = true;
+                    LogManager.GetLogger("testlogger").Info("Favorite {Planet}", new { Name = "Earth", Galaxy = "Milky way", Colors = new[] { "Red", "Green", "Blue" }, PlanetType = PlanetType.Ocean });
+                    return Task.FromResult(0);
+                }, includeEventProperties: true, configFn: googleTarget => googleTarget.SendJsonPayload = true);
+            Assert.Single(uploadedEntries);
+            var entry0 = uploadedEntries[0];
+            Assert.Equal("", entry0.TextPayload?.Trim() ?? "");
+            Assert.Equal(1, entry0.JsonPayload.Fields["properties"].StructValue.Fields.Count);
+            Assert.Equal(4, entry0.JsonPayload.Fields["properties"].StructValue.Fields["Planet"].StructValue.Fields.Count);
```
Comment (chrisdunelm) [inline apis/Google.Cloud.Logging.NLog/Google.Cloud.Logging.NLog.Tests/GoogleStackdriverTargetTest.cs]: nit: Please use some intermediate variables so the following lines aren't so long and repetitive.

File: apis/Google.Cloud.Logging.NLog/Google.Cloud.Logging.NLog/GoogleStackdriverTarget.cs [hunk 2]
```diff
@@ -127,6 +127,52 @@ protected override void InitializeTarget()
             base.InitializeTarget();
         }
 
+        private Func<object, Value> BuildProtoConverter()
+        {
+            // Create reusable JsonSerializer to reduce allocations
+            var jsonSettings = new Newtonsoft.Json.JsonSerializerSettings() { ReferenceLoopHandling = Newtonsoft.Json.ReferenceLoopHandling.Ignore };
```
Comment (chrisdunelm) [inline apis/Google.Cloud.Logging.NLog/Google.Cloud.Logging.NLog/GoogleStackdriverTarget.cs]: nit: Please format with initializer on its own line
``` C#
var jsonSettings = new Newtonsoft.Json.JsonSerializerSettings
{
    ReferenceLoopHandling = Newtonsoft.Json.ReferenceLoopHandling.Ignore
}
```

File: apis/Google.Cloud.Logging.NLog/Google.Cloud.Logging.NLog/GoogleStackdriverTarget.cs [hunk 3]
```diff
@@ -127,6 +127,52 @@ protected override void InitializeTarget()
             base.InitializeTarget();
         }
 
+        private Func<object, Value> BuildProtoConverter()
+        {
+            // Create reusable JsonSerializer to reduce allocations
+            var jsonSettings = new Newtonsoft.Json.JsonSerializerSettings() { ReferenceLoopHandling = Newtonsoft.Json.ReferenceLoopHandling.Ignore };
+            jsonSettings.Converters.Add(new Newtonsoft.Json.Converters.StringEnumConverter());
+            var jsonSerializer = Newtonsoft.Json.JsonSerializer.CreateDefault(jsonSettings);
+            return new Func<object, Value>(o => {
```
Comment (chrisdunelm) [inline apis/Google.Cloud.Logging.NLog/Google.Cloud.Logging.NLog/GoogleStackdriverTarget.cs]: Please use the simpler syntax with the open-brace on its own line:
``` C#
return o =>
{
    ...
};
```

File: apis/Google.Cloud.Logging.NLog/Google.Cloud.Logging.NLog.Tests/GoogleStackdriverTargetTest.cs (hunk 1 above)
Comment (snakefoot) [reply apis/Google.Cloud.Logging.NLog/Google.Cloud.Logging.NLog.Tests/GoogleStackdriverTargetTest.cs]: Fixed

File: apis/Google.Cloud.Logging.NLog/Google.Cloud.Logging.NLog/GoogleStackdriverTarget.cs (hunk 2 above)
Comment (snakefoot) [reply apis/Google.Cloud.Logging.NLog/Google.Cloud.Logging.NLog/GoogleStackdriverTarget.cs]: Fixed

File: apis/Google.Cloud.Logging.NLog/Google.Cloud.Logging.NLog/GoogleStackdriverTarget.cs (hunk 3 above)
Comment (snakefoot) [reply apis/Google.Cloud.Logging.NLog/Google.Cloud.Logging.NLog/GoogleStackdriverTarget.cs]: Fixed

**Comment by snakefoot** [comment]
@chrisdunelm Have updated the PR according to your review comments.

**Pull request merged**